dvc/repo/diff.py

```diff
@@ -190,29 +192,23 @@ def _targets_to_path_infos(repo_fs, targets):
 
 
 def _calculate_renamed(new, old, added, deleted):
-    old_inverted = {}
+    old_inverted: Dict[str, List[str]] = defaultdict(list)
     # It is needed to be dict of lists to cover cases
     # when repo has paths with same hash
     for path, path_hash in old.items():
-        bucket = old_inverted.get(path_hash, None)
-        if bucket is None:
-            old_inverted[path_hash] = [path]
-        else:
-            bucket.append(path)
+        old_inverted[path_hash].append(path)
 
     renamed = []
-    for path in added:
+    for path in sorted(added, key=len):
         path_hash = new[path]
-        old_paths = old_inverted.get(path_hash, None)
-        if not old_paths:
-            continue
-        old_path = None
-        for tmp_old_path in old_paths:
-            if tmp_old_path in deleted:
-                old_path = tmp_old_path
-                break
-        if not old_path:
+        old_paths = old_inverted[path_hash]
+        try:
+            iterator = enumerate(old_paths)
+            index = next(idx for idx, path in iterator if path in deleted)
+        except StopIteration:
             continue
+
+        old_path = old_paths.pop(index)
```

Comment:
So, in the case of multiple files with the same hashes, the only first file was being used as an old file.

To figure out renames,
1. we compare newly added and old deleted files and find renamed files.
2. then, we remove those that are renamed entries from `added` section using new path.
3. then, we remove those that are renamed entries from `deleted` section using old path.

We can have the same hashes for multiple renamed entries, but here in 1st step, we were only using the first encountered one. But, in the 3rd step, we delete it from `deleted` entries which fail if there were multiple files with the same hashes.